src/main/java/org/jenkinsci/plugins/workflow/support/storage/BulkFlowNodeStorage.java

```diff
@@ -0,0 +1,247 @@
+import com.thoughtworks.xstream.io.HierarchicalStreamReader;
+import com.thoughtworks.xstream.io.HierarchicalStreamWriter;
+import hudson.Util;
+import hudson.model.Action;
+import hudson.util.IOUtils;
+import hudson.util.RobustReflectionConverter;
+import hudson.util.XStream2;
+import org.jenkinsci.plugins.workflow.actions.FlowNodeAction;
+import org.jenkinsci.plugins.workflow.flow.FlowExecution;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.flow.FlowDurabilityHint;
+import org.jenkinsci.plugins.workflow.support.PipelineIOUtils;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.nio.file.Files;
+import java.nio.file.StandardOpenOption;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * {@link FlowNodeStorage} implementation that stores all the {@link FlowNode}s together in one file for efficient bulk I/O
+ *
+ * <p>This defers persisting until {@link #flush()} is called (or until we flush individual nodes explicitly or by
+ *  storing them without specifying delayWritingActions=true. It also doesn't use the atomic write operations.
+ *
+ *  Performance characteristics: much better use of the filesystem and far more efficient read/write if you do it all at once.
+ *  HOWEVER, if you insist on explicitly writing out each node, this reverts to overall O(n^2) performance, where n is node count.
+ *
+ * For these reasons, this implementation should <strong>only</strong> be used where {@link FlowDurabilityHint#isPersistWithEveryStep()}
+ * is <strong>false</strong>.
+ */
+public class BulkFlowNodeStorage extends FlowNodeStorage {
+    private final File dir;
+
+    private final FlowExecution exec;
+
+    /** Lazy-loaded mapping. */
+    private transient HashMap<String, Tag> nodes = null;
+
+    /** If true, we've been modified since last flush. */
+    private boolean isModified = false;
+
+    File getStoreFile() throws IOException {
+        return new File(dir, "flowNodeStore.xml");
+    }
+
+    public BulkFlowNodeStorage(FlowExecution exec, File dir) {
+        this.exec = exec;
+        this.dir = dir;
+        this.nodes = null;
+        this.setAvoidAtomicWrite(true);
+    }
+
+    /** Loads the nodes listing, lazily - so loading the {@link FlowExecution} doesn't trigger a more complex load. */
+    HashMap<String, Tag> getOrLoadNodes() throws IOException {
+        if (nodes == null) {
+            if (dir.exists()) {
+                File storeFile = getStoreFile();
+                if (storeFile.exists()) {
+                    HashMap<String, Tag> roughNodes = (HashMap<String, Tag>) (XSTREAM.fromXML(getStoreFile()));
```

Comment:
Redundant parens?

Reply:
Nah, I just like to enclose things I'm casting to help visually separate the thing being cast.

Reply:
Fair enough!